File: test/os_process.c
/* Behavioral contract for the process control kernels and host operations, run
 * against whichever implementation the build selected (`-Dos-process=c` or the
 * Zig default).
 *
 * The kernel section calls the seam directly. Command-line escaping has no
 * caller outside Windows, the signal lookup reports a position that
 * `os/proc-kill` turns into a panic, and the wait classification is collapsed
 * into a single number before Janet sees it, so none of the three can be
 * observed from the public functions alone. The public section then pins what
 * a caller does see, including the misspelled signal keyword recorded in
 * FOUND.md. */

#include <assert.h>
#include <stdio.h>
#include <string.h>
#include <janet.h>

#ifndef JANET_WINDOWS
#include <signal.h>
#include <unistd.h>
#endif

#define JANET_OS_WAIT_EXITED 0
#define JANET_OS_WAIT_STOPPED 1
#define JANET_OS_WAIT_SIGNALED 2
#define JANET_OS_WAIT_UNKNOWN 3

int32_t janet_os_exec_escape_arg(const char *arg, uint8_t *dest, int32_t cap);
int32_t janet_os_env_key_ok(const uint8_t *key, int32_t len);
void janet_os_env_entry_fill(const uint8_t *key, int32_t klen, const uint8_t *value,
                             int32_t vlen, uint8_t *dest);
int64_t janet_os_getpid(void);
int32_t janet_os_system(const char *command);

#ifndef JANET_WINDOWS
int32_t janet_os_signal_index(const uint8_t *key, int32_t len);
int32_t janet_os_wait(int64_t pid, int32_t *value);
void janet_os_reap(int64_t pid);
int32_t janet_os_kill(int64_t pid, int32_t sig);
int32_t janet_os_pipe(int *fds);
int32_t janet_os_close_fd(int fd);
int64_t janet_os_fork(void);
int32_t janet_os_exec(const char *path, char *const *argv, int32_t search_path);
#endif

/* Escaping is measured with nowhere to write, then filled into exactly the
 * space it asked for. Both passes must agree, and neither may touch a byte past
 * the reported length. */
static void expect_escape(const char *arg, const char *expected) {
    uint8_t buffer[64];
    int32_t needed = janet_os_exec_escape_arg(arg, NULL, 0);
    assert(needed == (int32_t) strlen(expected));
    assert(needed < (int32_t) sizeof(buffer));
    memset(buffer, '#', sizeof(buffer));
    assert(janet_os_exec_escape_arg(arg, buffer, needed) == needed);
    assert(memcmp(buffer, expected, (size_t) needed) == 0);
    assert(buffer[needed] == '#');
}

static void test_exec_escaping(void) {
    /* An argument the splitter would keep whole is passed through. */
    expect_escape("simple", "simple");
    expect_escape("", "");
    expect_escape("a\\b\\c", "a\\b\\c");
    expect_escape("trailing\\\\", "trailing\\\\");

    /* Each byte the splitter treats as a separator forces quoting. */
    expect_escape("two words", "\"two words\"");
    expect_escape("tab\there", "\"tab\there\"");
    expect_escape("line\nbreak", "\"line\nbreak\"");
    expect_escape("vertical\vtab", "\"vertical\vtab\"");

    /* A quotation mark is escaped, and so is every backslash before it. */
    expect_escape("a\"b", "\"a\\\"b\"");
    expect_escape("\"", "\"\\\"\"");
    expect_escape("a\\\\\"b", "\"a\\\\\\\\\\\"b\"");

    /* Backslashes elsewhere in a quoted argument stand for themselves. */
    expect_escape("a\\\\b c", "\"a\\\\b c\"");

    /* A run at the end meets the closing mark, so it doubles as well. */
    expect_escape("end \\\\", "\"end \\\\\\\\\"");

    /* Writing is refused when the caller offers less room than it asked for,
     * but the length is still reported. */
    uint8_t small[4];
    memset(small, '#', sizeof(small));
    assert(janet_os_exec_escape_arg("two words", small, 0) == 11);
    assert(small[0] == '#');
}

static void test_env_entries(void) {
    /* A key holding a separator or a terminator would be read back as a
     * different name, so it is refused. */
    assert(janet_os_env_key_ok((const uint8_t *) "PATH", 4) == 1);
    assert(janet_os_env_key_ok((const uint8_t *) "", 0) == 1);
    assert(janet_os_env_key_ok((const uint8_t *) "A=B", 3) == 0);
    assert(janet_os_env_key_ok((const uint8_t *) "A\0B", 3) == 0);

    /* Only the reported length is examined, so a separator past it is unseen. */
    assert(janet_os_env_key_ok((const uint8_t *) "A=B", 1) == 1);

    uint8_t entry[16];
    memset(entry, '#', sizeof(entry));
    janet_os_env_entry_fill((const uint8_t *) "K", 1, (const uint8_t *) "V", 1, entry);
    assert(memcmp(entry, "K=V", 4) == 0);
    assert(entry[4] == '#');

    janet_os_env_entry_fill((const uint8_t *) "K", 1, (const uint8_t *) "", 0, entry);
    assert(memcmp(entry, "K=", 3) == 0);

    /* A value may hold anything the key may not. */
    janet_os_env_entry_fill((const uint8_t *) "K", 1, (const uint8_t *) "a=b", 3, entry);
    assert(memcmp(entry, "K=a=b", 6) == 0);
}

#ifndef JANET_WINDOWS

static void test_signal_lookup(void) {
    assert(janet_os_signal_index((const uint8_t *) "kill", 4) == 0);
    assert(janet_os_signal_index((const uint8_t *) "int", 3) == 1);
    assert(janet_os_signal_index((const uint8_t *) "segv", 4) == 5);
    assert(janet_os_signal_index((const uint8_t *) "xfsz", 4) == 27);

    /* Only whole names match. */
    assert(janet_os_signal_index((const uint8_t *) "kil", 3) == -1);
    assert(janet_os_signal_index((const uint8_t *) "killer", 6) == -1);
    assert(janet_os_signal_index((const uint8_t *) "", 0) == -1);
    assert(janet_os_signal_index((const uint8_t *) "nosuch", 6) == -1);

    /* A key whose own bytes end in a terminator matches the shorter name, which
     * is what janet_cstrcmp did here. */
    assert(janet_os_signal_index((const uint8_t *) "int\0x", 5) == 1);

    /* The table misspells the name SIGVTALRM would give, so the documented
     * spelling is the one that fails. See FOUND.md. */
    assert(janet_os_signal_index((const uint8_t *) "vtlarm", 6) == 25);
    assert(janet_os_signal_index((const uint8_t *) "vtalrm", 6) == -1);
}

static void test_host_operations(void) {
    assert(janet_os_getpid() > 0);
    assert(janet_os_system("exit 0") == 0);
    assert(janet_os_system("exit 5") != 0);

    /* A pipe carries bytes from its write end to its read end. */
    int fds[2];
    char byte = 0;
    assert(janet_os_pipe(fds) == 0);
    assert(write(fds[1], "x", 1) == 1);
    assert(read(fds[0], &byte, 1) == 1);
    assert(byte == 'x');
    assert(janet_os_close_fd(fds[0]) == 0);
    assert(janet_os_close_fd(fds[1]) == 0);

    /* An ordinary exit is classified by its code. */
    int32_t value = -1;
    int64_t pid = janet_os_fork();
    assert(pid >= 0);
    if (pid == 0) _exit(3);
    assert(janet_os_wait(pid, &value) == JANET_OS_WAIT_EXITED);
    assert(value == 3);

    /* A process that ends on a signal is classified by that signal, without the
     * offset the caller adds. */
    value = -1;
    pid = janet_os_fork();
    assert(pid >= 0);
    if (pid == 0) {
        raise(SIGTERM);
        _exit(0);
    }
    assert(janet_os_wait(pid, &value) == JANET_OS_WAIT_SIGNALED);
    assert(value == SIGTERM);

    /* The same holds for a signal sent from outside. */
    value = -1;
    pid = janet_os_fork();
    assert(pid >= 0);
    if (pid == 0) {
        pause();
        _exit(0);
    }
    assert(janet_os_kill(pid, SIGKILL) == 0);
    assert(janet_os_wait(pid, &value) == JANET_OS_WAIT_SIGNALED);
    assert(value == SIGKILL);

    /* Collecting a process reports nothing, and a later wait on it finds
     * nothing to report. The untouched status word then classifies as a zero
     * exit, which is what the C implementation did by ignoring waitpid's
     * result. */
    pid = janet_os_fork();
    assert(pid >= 0);
    if (pid == 0) _exit(9);
    janet_os_reap(pid);
    value = -1;
    assert(janet_os_wait(pid, &value) == JANET_OS_WAIT_EXITED);
    assert(value == 0);

    /* Replacing a process by absolute path, and by a name found on the path. */
    value = -1;
    pid = janet_os_fork();
    assert(pid >= 0);
    if (pid == 0) {
        char *const args[] = {(char *) "/bin/sh", (char *) "-c", (char *) "exit 5", NULL};
        janet_os_exec("/bin/sh", args, 0);
        _exit(70);
    }
    assert(janet_os_wait(pid, &value) == JANET_OS_WAIT_EXITED);
    assert(value == 5);

    value = -1;
    pid = janet_os_fork();
    assert(pid >= 0);
    if (pid == 0) {
        char *const args[] = {(char *) "sh", (char *) "-c", (char *) "exit 6", NULL};
        janet_os_exec("sh", args, 1);
        _exit(70);
    }
    assert(janet_os_wait(pid, &value) == JANET_OS_WAIT_EXITED);
    assert(value == 6);

    /* A failed replacement returns instead of ending the process. */
    value = -1;
    pid = janet_os_fork();
    assert(pid >= 0);
    if (pid == 0) {
        char *const args[] = {(char *) "/janet-zig-os-process-absent-4f81", NULL};
        _exit(janet_os_exec("/janet-zig-os-process-absent-4f81", args, 0) == -1 ? 71 : 72);
    }
    assert(janet_os_wait(pid, &value) == JANET_OS_WAIT_EXITED);
    assert(value == 71);
}

/* Each contract runs as one function so that a form which waits on the event
 * loop finishes before the next contract starts. janet_dostring evaluates each
 * top level form in its own fiber and does not drive the loop, which would let
 * a spawn outlive the wait that follows it. */
static void run(JanetTable *env, const char *source) {
    char wrapped[4096];
    int written = snprintf(wrapped, sizeof(wrapped), "(fn [] %s)", source);
    assert(written > 0 && written < (int) sizeof(wrapped));

    Janet fnv = janet_wrap_nil();
    assert(janet_dostring(env, wrapped, "os-process-contract", &fnv) == 0);
    assert(janet_checktype(fnv, JANET_FUNCTION));

    JanetFiber *fiber = janet_fiber(janet_unwrap_function(fnv), 64, 0, NULL);
    fiber->env = env;
#ifdef JANET_EV
    janet_gcroot(janet_wrap_fiber(fiber));
    janet_schedule(fiber, janet_wrap_nil());
    janet_loop();
    assert(janet_fiber_status(fiber) == JANET_STATUS_DEAD);
    janet_gcunroot(janet_wrap_fiber(fiber));
#else
    Janet result = janet_wrap_nil();
    JanetSignal signal = janet_continue(fiber, janet_wrap_nil(), &result);
    if (signal != JANET_SIGNAL_OK) janet_stacktrace_ext(fiber, result, "");
    assert(signal == JANET_SIGNAL_OK);
#endif
}

static void test_core_functions(void) {
    JanetTable *env = janet_core_env(NULL);

    /* An exit code reaches the caller unchanged, whether the program is named
     * by path or found on it. */
    run(env,
        "(assert (= 0 (os/execute [\"/bin/sh\" \"-c\" \"exit 0\"])))\n"
        "(assert (= 7 (os/execute [\"/bin/sh\" \"-c\" \"exit 7\"])))\n"
        "(assert (= 3 (os/execute [\"sh\" \"-c\" \"exit 3\"] :p)))\n");

    /* The :x flag turns a non-zero code into an error — but only where the
     * event loop is compiled in. The flag is read in the wait callback, which
     * does not exist otherwise, so a build without the event loop accepts the
     * flag and ignores it. That is recorded in FOUND.md and left unfixed, so
     * the contract pins each configuration as it stands. */
    run(env,
        "(assert (= 0 (os/execute [\"/bin/sh\" \"-c\" \"exit 0\"] :x)))\n"
#ifdef JANET_EV
        "(assert (not (first (protect (os/execute [\"/bin/sh\" \"-c\" \"exit 1\"] :x)))))\n"
#else
        "(assert (= 1 (os/execute [\"/bin/sh\" \"-c\" \"exit 1\"] :x)))\n"
#endif
    );

    /* A supplied environment reaches the child, and a key holding a separator
     * is dropped rather than passed as a different name. */
    run(env,
        "(assert (= 0 (os/execute [\"/bin/sh\" \"-c\" \"[ \\\"$FOO\\\" = bar ]\"] :e"
        " {\"FOO\" \"bar\"})))\n"
        "(assert (= 0 (os/execute [\"/bin/sh\" \"-c\" \"[ -z \\\"$A\\\" ]\"] :e"
        " {\"A=B\" \"C\" \"FOO\" \"bar\"})))\n");

    /* A process ended by a signal reports that signal offset by 128. */
    run(env,
        "(def p (os/spawn [\"/bin/sh\" \"-c\" \"sleep 10\"]))\n"
        "(assert (= 143 (os/proc-kill p true :term)))\n");

    /* Signal keywords are looked up whole, and the table's misspelling of
     * SIGVTALRM is the spelling that resolves. See FOUND.md. */
    run(env,
        "(def p (os/spawn [\"/bin/sh\" \"-c\" \"sleep 10\"]))\n"
        "(assert (not (first (protect (os/proc-kill p false :vtalrm)))))\n"
        "(assert (not (first (protect (os/proc-kill p false :kil)))))\n"
        "(os/proc-kill p false :vtlarm)\n"
        "(assert (>= (os/proc-wait p) 129))\n");

    /* Waiting twice on the same process is an error. */
    run(env,
        "(def p (os/spawn [\"/bin/sh\" \"-c\" \"exit 4\"]))\n"
        "(assert (= 4 (os/proc-wait p)))\n"
        "(assert (not (first (protect (os/proc-wait p)))))\n");

    /* The remaining process functions report the host directly.
     *
     * os/shell is called without a command, which is the only form that
     * survives: passing one aborts the process under the event loop, because
     * the subroutine frees the copied command and the default callback frees
     * it again. That is recorded in FOUND.md and left unfixed, so the contract
     * cannot exercise it. */
    run(env,
        "(assert (> (os/getpid) 0))\n"
        "(assert (boolean? (os/shell)))\n");
}

#endif /* JANET_WINDOWS */

int main(void) {
    test_exec_escaping();
    test_env_entries();
#ifndef JANET_WINDOWS
    test_signal_lookup();
    test_host_operations();

    janet_init();
    test_core_functions();
    janet_deinit();
#endif
    return 0;
}

File: test/os_process.zig
//! Behavioral contract for the process-control kernels and the host calls
//! behind Janet's process functions.
//!
//! ## What only the kernels can be asked
//!
//! Three of the four sections below have no spelling in Janet at all.
//! Command-line escaping has no caller outside Windows, the signal lookup
//! reports a *position* that `os/proc-kill` turns into a number and then into
//! a panic, and the wait classification is collapsed into a single integer
//! before Janet sees it. The public section then pins what a caller does see.
//!
//! ## How the subjects are reached
//!
//! Every kernel here is reached by import, so nothing is declared twice.
//!
//! The four wait codes are named rather than restated. A third copy of a
//! number two files already agree on is a place they can drift;
//! `os_process.wait_exited` and its three siblings are the subject's own, so a
//! renumbering fails to compile here instead of passing against a stale
//! literal.
//!
//! ## Some of this is asserted twice, deliberately
//!
//! `os/process.zig` has `test` blocks over the escaping, the signal lookup and
//! the environment rule, and `zig build test` runs them. They are a different
//! instrument from this file rather than a duplicate of it, and both are kept.

// ==========================================================================
// Standard library imports
// ==========================================================================

const std = @import("std");
const builtin = @import("builtin");

// ==========================================================================
// Project imports
// ==========================================================================

const expect = @import("expect.zig").expect;
const harness = @import("harness.zig");
const os_process = subsystems.process;
const subsystems = @import("subsystems");
const vm_lifecycle = @import("subsystems").lifecycle;

// ==========================================================================
// Constants
// ==========================================================================

const windows = builtin.os.tag == .windows;

// ==========================================================================
// Types
// ==========================================================================

/// The host calls a contract needs and a subsystem does not: a child that
/// exits without flushing, a child that stops until it is signalled, and the
/// two ends of a pipe.
const posix = struct {
    extern fn write(fd: c_int, buffer: [*]const u8, count: usize) callconv(.c) isize;
    extern fn read(fd: c_int, buffer: [*]u8, count: usize) callconv(.c) isize;
    extern fn raise(sig: c_int) callconv(.c) c_int;

    /// `std.c.SIG` is an enum on some targets and a plain integer on others,
    /// so a signal number is narrowed once here rather than at each site.
    fn signal(number: anytype) c_int {
        return switch (@typeInfo(@TypeOf(number))) {
            .@"enum" => @intCast(@intFromEnum(number)),
            else => @intCast(number),
        };
    }
    extern fn pause() callconv(.c) c_int;
    extern fn _exit(status: c_int) callconv(.c) noreturn;
};

// ==========================================================================
// Cases
// ==========================================================================

/// Escaping is measured with nowhere to write, then filled into exactly the
/// space it asked for. Both passes must agree, and neither may touch a byte
/// past the reported length, so the destination is poisoned first and the byte
/// after the result is checked.
fn expectEscape(argument: [*:0]const u8, expected: []const u8) void {
    var buffer: [64]u8 = undefined;
    const needed = os_process.escapeArgument(argument, null, 0);
    expect(needed == @as(i32, @intCast(expected.len)));
    expect(needed < @as(i32, @intCast(buffer.len)));

    @memset(&buffer, '#');
    expect(os_process.escapeArgument(argument, &buffer, needed) == needed);
    expect(std.mem.eql(u8, buffer[0..@intCast(needed)], expected));
    expect(buffer[@intCast(needed)] == '#');
}

fn theExecEscaping() void {
    // An argument the splitter would keep whole is passed through.
    expectEscape("simple", "simple");
    expectEscape("", "");
    expectEscape("a\\b\\c", "a\\b\\c");
    expectEscape("trailing\\\\", "trailing\\\\");

    // Each byte the splitter treats as a separator forces quoting.
    expectEscape("two words", "\"two words\"");
    expectEscape("tab\there", "\"tab\there\"");
    expectEscape("line\nbreak", "\"line\nbreak\"");
    expectEscape("vertical\x0btab", "\"vertical\x0btab\"");

    // A quotation mark is escaped, and so is every backslash before it.
    expectEscape("a\"b", "\"a\\\"b\"");
    expectEscape("\"", "\"\\\"\"");
    expectEscape("a\\\\\"b", "\"a\\\\\\\\\\\"b\"");

    // Backslashes elsewhere in a quoted argument stand for themselves.
    expectEscape("a\\\\b c", "\"a\\\\b c\"");

    // A run at the end meets the closing mark, so it doubles as well.
    expectEscape("end \\\\", "\"end \\\\\\\\\"");

    // Writing is refused when the caller offers less room than it asked for,
    // but the length is still reported.
    var small: [4]u8 = undefined;
    @memset(&small, '#');
    expect(os_process.escapeArgument("two words", &small, 0) == 11);
    expect(small[0] == '#');
}

fn theEnvEntries() void {
    // A key containing a separator or a terminator would be read back as a
    // different name, so it is refused.
    expect(os_process.envKeyOk("PATH", 4) == 1);
    expect(os_process.envKeyOk("", 0) == 1);
    expect(os_process.envKeyOk("A=B", 3) == 0);
    expect(os_process.envKeyOk("A\x00B", 3) == 0);

    // Only the reported length is examined, so a separator past it is unseen.
    expect(os_process.envKeyOk("A=B", 1) == 1);

    var entry: [16]u8 = undefined;
    @memset(&entry, '#');
    os_process.envEntryFill("K", 1, "V", 1, &entry);
    expect(std.mem.eql(u8, entry[0..4], "K=V\x00"));
    expect(entry[4] == '#');

    os_process.envEntryFill("K", 1, "", 0, &entry);
    expect(std.mem.eql(u8, entry[0..3], "K=\x00"));

    // A value may contain anything the key may not.
    os_process.envEntryFill("K", 1, "a=b", 3, &entry);
    expect(std.mem.eql(u8, entry[0..6], "K=a=b\x00"));
}

fn theSignalLookup() void {
    expect(os_process.signalIndex("kill", 4) == 0);
    expect(os_process.signalIndex("int", 3) == 1);
    expect(os_process.signalIndex("segv", 4) == 5);
    expect(os_process.signalIndex("xfsz", 4) == 27);

    // Only whole names match.
    expect(os_process.signalIndex("kil", 3) == -1);
    expect(os_process.signalIndex("killer", 6) == -1);
    expect(os_process.signalIndex("", 0) == -1);
    expect(os_process.signalIndex("nosuch", 6) == -1);

    // A key whose own bytes end in a terminator matches the shorter name,
    // which is `utils.cstrcmp`'s rule.
    expect(os_process.signalIndex("int\x00x", 5) == 1);

    // Every name is its signal's own, lower-cased with the `SIG` dropped, and
    // `vtalrm` is no exception: `vtlarm` is not an alias for it.
    expect(os_process.signalIndex("vtalrm", 6) == 25);
    expect(os_process.signalIndex("vtlarm", 6) == -1);
}

fn theHostOperations() void {
    expect(os_process.processId() > 0);
    expect(os_process.shell("exit 0") == 0);
    expect(os_process.shell("exit 5") != 0);

    // A pipe moves bytes from its write end to its read end.
    var fds: [2]c_int = undefined;
    var byte: [1]u8 = .{0};
    expect(os_process.makePipe(&fds) == 0);
    expect(posix.write(fds[1], "x", 1) == 1);
    expect(posix.read(fds[0], &byte, 1) == 1);
    expect(byte[0] == 'x');
    expect(os_process.closeDescriptor(fds[0]) == 0);
    expect(os_process.closeDescriptor(fds[1]) == 0);

    // An ordinary exit is classified by its code.
    var value: i32 = -1;
    var pid = os_process.forkProcess();
    expect(pid >= 0);
    if (pid == 0) posix._exit(3);
    expect(os_process.wait(pid, &value) == os_process.wait_exited);
    expect(value == 3);

    // A process that ends on a signal is classified by that signal, without
    // the offset the caller adds.
    value = -1;
    pid = os_process.forkProcess();
    expect(pid >= 0);
    if (pid == 0) {
        _ = posix.raise(posix.signal(std.c.SIG.TERM));
        posix._exit(0);
    }
    expect(os_process.wait(pid, &value) == os_process.wait_signaled);
    expect(value == posix.signal(std.c.SIG.TERM));

    // The same is true of a signal sent from outside.
    value = -1;
    pid = os_process.forkProcess();
    expect(pid >= 0);
    if (pid == 0) {
        _ = posix.pause();
        posix._exit(0);
    }
    expect(os_process.sendSignal(pid, posix.signal(std.c.SIG.KILL)) == 0);
    expect(os_process.wait(pid, &value) == os_process.wait_signaled);
    expect(value == posix.signal(std.c.SIG.KILL));

    // Collecting a process reports nothing, and a later wait on it finds
    // nothing to report. The untouched status word then classifies as a zero
    // exit, which is what the C implementation did by ignoring `waitpid`'s
    // result.
    pid = os_process.forkProcess();
    expect(pid >= 0);
    if (pid == 0) posix._exit(9);
    os_process.reap(pid);
    value = -1;
    expect(os_process.wait(pid, &value) == os_process.wait_exited);
    expect(value == 0);

    // Replacing a process by absolute path, and by a name found on the path.
    value = -1;
    pid = os_process.forkProcess();
    expect(pid >= 0);
    if (pid == 0) {
        const args = [_:null]?[*:0]const u8{ "/bin/sh", "-c", "exit 5" };
        _ = os_process.exec("/bin/sh", &args, 0);
        posix._exit(70);
    }
    expect(os_process.wait(pid, &value) == os_process.wait_exited);
    expect(value == 5);

    value = -1;
    pid = os_process.forkProcess();
    expect(pid >= 0);
    if (pid == 0) {
        const args = [_:null]?[*:0]const u8{ "sh", "-c", "exit 6" };
        _ = os_process.exec("sh", &args, 1);
        posix._exit(70);
    }
    expect(os_process.wait(pid, &value) == os_process.wait_exited);
    expect(value == 6);

    // A failed replacement returns instead of ending the process.
    value = -1;
    pid = os_process.forkProcess();
    expect(pid >= 0);
    if (pid == 0) {
        const absent = "/janet-zig-os-process-absent-4f81";
        const args = [_:null]?[*:0]const u8{absent};
        posix._exit(if (os_process.exec(absent, &args, 0) == -1) 71 else 72);
    }
    expect(os_process.wait(pid, &value) == os_process.wait_exited);
    expect(value == 71);
}

fn theCoreFunctions() void {
    const env = harness.coreEnv();

    // An exit code reaches the caller unchanged, whether the program is named
    // by path or found on it.
    harness.inFiber(env,
        \\(assert (= 0 (os/execute ["/bin/sh" "-c" "exit 0"])))
        \\(assert (= 7 (os/execute ["/bin/sh" "-c" "exit 7"])))
        \\(assert (= 3 (os/execute ["sh" "-c" "exit 3"] :p)))
    );

    // The `:x` flag turns a non-zero code into an error in every
    // configuration. The flag is read in the evented wait callback, which
    // does not exist without the event loop, so the non-evented wait reads it
    // too rather than accepting the flag and ignoring it; a caller relying
    // on `:x` for error handling would otherwise get none, silently, in one
    // build and not the other. Both raise the same message.
    harness.inFiber(env,
        \\(assert (= 0 (os/execute ["/bin/sh" "-c" "exit 0"] :x)))
        \\(assert (= "command failed with non-zero exit code 1"
        \\           (in (protect (os/execute ["/bin/sh" "-c" "exit 1"] :x)) 1)))
    );

    // A supplied environment reaches the child, and a key with a separator in
    // it
    // is dropped rather than passed as a different name.
    harness.inFiber(env,
        \\(assert (= 0 (os/execute ["/bin/sh" "-c" "[ \"$FOO\" = bar ]"] :e {"FOO" "bar"})))
        \\(assert (= 0 (os/execute ["/bin/sh" "-c" "[ -z \"$A\" ]"] :e {"A=B" "C" "FOO" "bar"})))
    );

    // A process ended by a signal reports that signal offset by 128.
    harness.inFiber(env,
        \\(def p (os/spawn ["/bin/sh" "-c" "sleep 10"]))
        \\(assert (= 143 (os/proc-kill p true :term)))
    );

    // Signal keywords are looked up whole, and `vtalrm` is the spelling that
    // resolves: the signal's own name with the `SIG` dropped, like every
    // other row of the table. A prefix of a name is not a name.
    harness.inFiber(env,
        \\(def p (os/spawn ["/bin/sh" "-c" "sleep 10"]))
        \\(assert (not (first (protect (os/proc-kill p false :vtlarm)))))
        \\(assert (not (first (protect (os/proc-kill p false :kil)))))
        \\(os/proc-kill p false :vtalrm)
        \\(assert (>= (os/proc-wait p) 129))
    );

    // Waiting twice on the same process is an error.
    harness.inFiber(env,
        \\(def p (os/spawn ["/bin/sh" "-c" "exit 4"]))
        \\(assert (= 4 (os/proc-wait p)))
        \\(assert (not (first (protect (os/proc-wait p)))))
    );

    // The remaining process functions report the host directly.
    //
    // `os/shell` is called with a command as well as without one. The command
    // form is the one with a subroutine under it: that subroutine frees the
    // copy it was given, and the reply is a tag with no payload, so a callback
    // that freed for every tag would free it a second time.
    harness.inFiber(env,
        \\(assert (> (os/getpid) 0))
        \\(assert (boolean? (os/shell)))
        \\(assert (= 0 (os/shell "exit 0")))
        \\# The answer is `system`'s wait status rather than the exit code,
        \\# which is what "pass a command string directly to the system shell"
        \\# means and is unchanged by the fix.
        \\(assert (= 768 (os/shell "exit 3")))
    );
}

// ==========================================================================
// Entry
// ==========================================================================

pub fn run() void {
    theExecEscaping();
    theEnvEntries();

    if (!windows) {
        theSignalLookup();
        theHostOperations();

        harness.init();
        theCoreFunctions();
        vm_lifecycle.deinit();
    }

    std.debug.print("os_process contract ok\n", .{});
}
